src/Mvc/Mvc.Core/src/Authorization/AuthorizeFilter.cs

```diff
@@ -207,15 +204,24 @@ public virtual async Task OnAuthorizationAsync(AuthorizationFilterContext contex
 
     IFilterMetadata IFilterFactory.CreateInstance(IServiceProvider serviceProvider)
     {
-        if (Policy != null || PolicyProvider != null)
+        if (Policy is null && PolicyProvider is null)
         {
-            // The filter is fully constructed. Use the current instance to authorize.
-            return this;
+            // REVIEW: Can serviceProvider ever change here? Can this ever be called in parallel?
```

Comment:
@pranavkm @javiercn

This method possibly being called with a different `serviceProvider` on the same `AuthorizeFilter` instance is what concerns me the most about this change. It doesn't seem super likely to me, but it also doesn't sound impossible. If possible, this could lead to some very-hard-to-debug behavior.

Parallel calls by themselves shouldn't be too big an issue if the `serviceProvider` remains the same since it would just be redoing work. I think this is generally only called once during startup though.

I think we'll just have to document the behavior and tell people to implement `IFilterFactory.CreateInstance(IServiceProvider serviceProvider)` themselves if they want to override `OnAuthorizationAsync`.